conventions/c++.rst

```diff
@@ -0,0 +1,414 @@
+  len);``.
+
+* It's generally better to build up strings using ``+=`` rather than
+  combining series of components with ``+``.  So ``foo = a + " and " +
+  c`` is better written as ``foo = a; foo += " and "; foo += c;``.
+  It's possible for compilers to handle the former without a lot of
+  temporary string objects by returning a proxy object to allow the
+  concatenation to happen lazily, but not all compilers do this, and
+  it's likely to still have some overhead.  Note that GCC 4.1 seems to
+  produce larger code in some cases for the latter approach, but it's
+  a definite win with GCC 4.4.
+
+* ``std::string(1, '\0')`` seems to be slightly more efficient than
+  ``std::string("", 1)`` for constructing a ``std::string`` containing
+  a single ASCII nul character.
+
+
+Use of C++ Features
+~~~~~~~~~~~~~~~~~~~
+
+C++11
+-----
+
+As of Xapian 1.3.3, a compiler with decent support for C++11 is
+required to build Xapian.  We currently aim to allow users to use a
+non-C++11 compiler to build code which uses Xapian.
+
+There are now several compilers with good C++11 support, but there are
+a few shortfalls in commonly deployed versions of most of them.  Often
+we can work around this, and we should do where the effort is low
+compared to the gain (so a compiler version which is widely used is
+more worth supporting than one which is hardly used by anyone).
+
+However, we shouldn't have to jump through hoops to cater for
+compilers where their authors aren't putting in the effort to keep up
+with the language standards.
+
+Please avoid the following C++11 features for the time being:
+
+* ``std::to_string()`` - this is completely missing on current
+  versions of mingw and cygwin - in the library, you can ``#include
+  "str.h"`` and then use the ``str()`` function instead for most
+  cases.  This is also usually faster than ``std::to_string()``.
+
+
+C++ features we assume
+----------------------
+
+We assume that all compilers will correctly implement the following,
+so it's safe to rely on them when working on Xapian.
+
+* We assume ``<sstream>`` is available.  GCC < 2.95.3 didn't have it
+  but GCC 2.95.3 includes a backported version.  We aren't aware of
+  any other compilers still in use which lack it.
+
+* Non-".h" versions of standard ISO C++ headers (e.g. ``#include
+  <list>`` rather than ``#include <list.h>``).  We aren't aware of any
+  compiler still in use which lacks these, and GCC 4.3 no longer has
+  the old versions.  If there are any, we could add a directory full
+  of forwarding headers to work around this issue.
+
+* Standard header ``<limits>`` (for ``numeric_limits<>``) - for GCC, this was
+  added in GCC 3.0.
+
+* Standard header ``<streambuf>`` (GCC < 3.0 only has ``<streambuf.h>``).
+
+
+Exceptions
+~~~~~~~~~~
+
+When catching an exception which is an object, do it by const reference, so
+like this::
+
+      try {
+	  foo();
+      } catch (const ErrorClass &e) {
```

Comment:
We're now preferring putting what seems to be generally preferred C++ style of putting the `&` or `*` with the typename, i.e. `ErrorClass&`.